homeassistant/helpers/entity_component.py

```diff
@@ -391,47 +357,117 @@ def async_add_entities(self, new_entities, update_before_add=False):
+        if update_before_add:
+            try:
+                yield from entity.async_device_update(warning=False)
+            except Exception:  # pylint: disable=broad-except
+                self.component.logger.exception(
+                    "%s: Error on device update!", self.platform)
+                return
+
+        # Write entity_id to entity
+        if getattr(entity, 'entity_id', None) is None:
+            object_id = entity.name or DEVICE_DEFAULT_NAME
+
+            if self.entity_namespace is not None:
+                object_id = '{} {}'.format(self.entity_namespace,
+                                           object_id)
+
+            entity.entity_id = async_generate_entity_id(
+                self.component.entity_id_format, object_id,
+                component_entities)
+
+        # Make sure it is valid in case an entity set the value themselves
+        if not valid_entity_id(entity.entity_id):
+            raise HomeAssistantError(
+                'Invalid entity id: {}'.format(entity.entity_id))
+        elif entity.entity_id in component_entities:
+            raise HomeAssistantError(
+                'Entity id already exists: {}'.format(entity.entity_id))
+
+        self.entities[entity.entity_id] = entity
+        component_entities.add(entity.entity_id)
+
+        if hasattr(entity, 'async_added_to_hass'):
+            yield from entity.async_added_to_hass()
+
+        yield from entity.async_update_ha_state()
+
     @asyncio.coroutine
     def async_reset(self):
         """Remove all entities and reset data.
 
         This method must be run in the event loop.
         """
-        if not self.platform_entities:
+        if not self.entities:
             return
 
-        tasks = [entity.async_remove() for entity in self.platform_entities]
+        tasks = [self._async_remove_entity(entity_id)
+                 for entity_id in self.entities]
 
         yield from asyncio.wait(tasks, loop=self.component.hass.loop)
 
         if self._async_unsub_polling is not None:
             self._async_unsub_polling()
             self._async_unsub_polling = None
 
+    @asyncio.coroutine
+    def async_remove_entity(self, entity_id):
+        """Remove entity id from platform."""
+        yield from self._async_remove_entity(entity_id)
+
+        # Clean up polling job if no longer needed
+        if (self._async_unsub_polling is not None and
+                not any(entity.should_poll for entity
+                        in self.entities.values())):
+            self._async_unsub_polling()
+            self._async_unsub_polling = None
+
+    @asyncio.coroutine
+    def _async_remove_entity(self, entity_id):
+        """Remove entity id from platform."""
+        entity = self.entities.pop(entity_id)
+
+        if hasattr(entity, 'async_will_remove_from_hass'):
+            yield from entity.async_will_remove_from_hass()
+
+        self.component.hass.states.async_remove(entity_id)
```

Comment:
I think we should not remove the old `async_remove` from entity. We can call it on this please instead to make a new hidden function. for `async_update` and maybe `async_added_to_hass` is that okay but I prefere (for software designe) a overwritte able exists function on base class that every one can see it and can overwrite it (maybe with pass).

Reply:
The reason I removed `Entity.async_remove` is because the platform is responsible for managing the entities of a platform. The entity cannot just decide to skip out without notifying the platform.

To help entities be notified of being removed, I've added the `async_will_remove_from_hass` hook. (similar to `async_added_to_hass`)

Reply:
btw, I am not against reinstating `Entity.async_remove`. It would look like this: https://github.com/home-assistant/home-assistant/pull/11691#issuecomment-358082013

Reply:
Maybe `async_remove_from_hass`. And we should add this method with pass to entity for the design. I think `async_update` should be the only function for look at it. This help also on async to can make a flip to other task and speed up the parallel. I don't think we lost some things with make coroutine with pass and it make it easyer to read. 

But I'm fine with this change but the #11691 is bad.

Reply:
I change my mind. I think #11691 is a good solution 👍